crates/ide/src/hover/render.rs

```diff
@@ -1210,3 +1222,85 @@ fn render_dyn_compatibility(
         }
     }
 }
+
+fn is_pwr2(val: u128) -> bool {
+    val.count_ones() == 1
+}
+
+fn is_pwr2minus1(val: u128) -> bool {
+    val == u128::MAX || (val + 1).count_ones() == 1
+}
+
+fn is_pwr2plus1(val: u128) -> bool {
+    val != 0 && (val - 1).count_ones() == 1
+}
+
+fn pwr2_to_exponent(num: u128) -> String {
+    const DIGITS: [char; 10] = ['⁰', '¹', '²', '³', '⁴', '⁵', '⁶', '⁷', '⁸', '⁹'];
+    (127 - num.leading_zeros())
```

Comment:
Wouldn't `num.trailing_zeros()` work here?

Reply:
There's a bunch of ways around that, like restructuring the code by splitting off "number to superscript digits" from "checking if it's a power of 2, more or less" so it's clearly correct, or by using an `assert!`, but if we have a bug here, it's better to make it more obvious anyway.

Reply:
Can do!

Reply:
Sorted in https://github.com/rust-lang/rust-analyzer/pull/18973/commits/705eb84941c2d2eb3daa741fb6a99216572737f4